storage/qingstor.go

```diff
@@ -0,0 +1,192 @@
+package storage
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/pengsrc/go-shared/convert"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"github.com/yunify/qingstor-sdk-go/v3/config"
+	"github.com/yunify/qingstor-sdk-go/v3/request/errors"
+	"github.com/yunify/qingstor-sdk-go/v3/service"
+
+	"github.com/yunify/qsctl/constants"
+)
+
+// QingStorObjectStorage will implement ObjectStorage interface.
+type QingStorObjectStorage struct {
+	service *service.Service
+	bucket  *service.Bucket
+}
+
+// NewQingStorObjectStorage will create a new qingstor object storage.
+func NewQingStorObjectStorage() (q *QingStorObjectStorage, err error) {
+	cfg, err := config.New(
+		viper.GetString(constants.ConfigAccessKeyID),
+		viper.GetString(constants.ConfigSecretAccessKey),
+	)
+	if err != nil {
+		log.Errorf("contexts: Init config failed [%v]", err)
+		return
+	}
+
+	cfg.Host = viper.GetString(constants.ConfigHost)
+	cfg.Port = viper.GetInt(constants.ConfigPort)
+	cfg.Protocol = viper.GetString(constants.ConfigProtocol)
+	cfg.ConnectionRetries = viper.GetInt(constants.ConfigConnectionRetries)
+	cfg.LogLevel = viper.GetString(constants.ConfigLogLevel)
+
+	q = &QingStorObjectStorage{}
+
+	q.service, err = service.Init(cfg)
+	if err != nil {
+		log.Errorf("contexts: Init service failed [%v]", err)
+		return
+	}
+	return
+}
+
+// SetupBucket implements ObjectStorage.SetupBucket
+func (q *QingStorObjectStorage) SetupBucket(name, zone string) (err error) {
+	if zone == "" {
+		zone = viper.GetString(constants.ConfigZone)
+	}
+
+	if zone != "" {
+		q.bucket, _ = q.service.Bucket(name, zone)
```

Comment:
In my opinion, we should still keep the error assert, even the function always return nil, in case of predictable changes in the future.

Reply:
OK, let's do it.